hs_core/hydro_realtime_signal_processor.py

```diff
@@ -0,0 +1,96 @@
+from django.db import models
+from haystack.signals import RealtimeSignalProcessor
+from haystack.exceptions import NotHandled
+import logging
+import types
+from haystack.query import SearchQuerySet
+from haystack.utils import get_identifier
+
+logger = logging.getLogger(__name__)
+
+
+class HydroRealtimeSignalProcessor(RealtimeSignalProcessor):
+
+    """ 
+    Customized for the fact that all indexed resources are subclasses of BaseResource. 
+
+    Notes: 
+    1. RealtimeSignalProcessor already plumbs in all class updates. We might want to be more specific. 
+    2. The class sent to this is a subclass of BaseResource, or another class. 
+    3. Thus, we want to capture cases in which it is an appropriate instance, and respond. 
+    """
+
+    def hydro_resource_is_present(self, instance): 
+        """ 
+        This returns True if a current resource is now present in Haystack, and False if not. 
+        This -- in turn -- informs us as to whether to update and/or delete the object 
+        """
+        discoverable = SearchQuerySet().filter(id=get_identifier(instance)).count()
+        if discoverable > 0:
+            return True
+        else:
+            return False
+
+    def handle_save(self, sender, instance, **kwargs):
+        """
+        Given an individual model instance, determine which backends the
+        update should be sent to & update the object on those backends.
+        """
+        from hs_core.models import BaseResource
+        from hs_access_control.models import ResourceAccess
+    
+        if isinstance(instance, BaseResource):
+            if hasattr(instance, 'raccess') and hasattr(instance, 'metadata'): 
+                # work around for failure of super(BaseResource, instance) to work properly.
+                # this always succeeds because this is a post-save object action. 
+                newinstance = BaseResource.objects.get(pk=instance.pk)
+                newsender = BaseResource
+                using_backends = self.connection_router.for_write(instance=newinstance)
+                for using in using_backends:
+                    # if object is public/discoverable or becoming public/discoverable, index it 
+                    if instance.raccess.public or instance.raccess.discoverable: 
+                        try:
+                            index = self.connections[using].get_unified_index().get_index(newsender)
+                            index.update_object(newinstance, using=using)
+                        except NotHandled:
+                            logger.exception("Failure: changes to " + str(type(instance)) + " with short_id " + newinstance.short_id+ " not added to SOLR index")
```

Comment:
Logging calls have string formatting built in, e.g. `logger.exception("Failure: changes to %s with short_id %s not added to Solr Index.", type(instance), newinstance.short_id)`, with the ability to add arbitrary numbers of positional arguments.

In general, building up strings using string formatting (either old-style, `"%s is a thing." % my_obj`, or new-style, `"{0} is a thing.".format(my_obj)`) is considered more "Pythonic" than using string concatenation.